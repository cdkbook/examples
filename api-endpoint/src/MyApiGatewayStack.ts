import { CfnOutput, Stack, StackProps } from 'aws-cdk-lib';
import { LambdaIntegration, RestApi } from 'aws-cdk-lib/aws-apigateway';
import { Certificate } from 'aws-cdk-lib/aws-certificatemanager';
import { ARecord, HostedZone, RecordTarget } from 'aws-cdk-lib/aws-route53';
import { ApiGateway } from 'aws-cdk-lib/aws-route53-targets';
import { Construct } from 'constructs';
import { NewsletterFunction } from './handlers/newsletter/newsletter-function';
import { ReviewsHandlerFunction } from './handlers/reviews/reviews-handler-function';

export interface MyApiGatewayStackProps extends StackProps {
  /**
   * The ARN of the ACM certificate to use for the endpoint.
   */
  readonly certificateArn: string;

  /**
   * The Hosted Zone ID of the domain being used.
   */
  readonly hostedZoneId: string;
}

export class MyApiGatewayStack extends Stack {
  constructor(scope: Construct, id: string, props: MyApiGatewayStackProps) {
    super(scope, id, props);

    const api = new RestApi(this, 'Api', {
      description: 'A simple API example for The CDK Book',
      restApiName: 'TheCDKBookApi',
      deploy: true, // let's make this easier by always redeploying,
      deployOptions: {
        stageName: 'v1',
        throttlingRateLimit: 10,
      },
      domainName: {
        domainName: 'api.thecdkbook.com',
        certificate: Certificate.fromCertificateArn(this, 'ApiCertificate', props.certificateArn),
      },
    });
    this.createNewsletterResource(api);
    this.createReviewsResource(api);
    this.addDns(api, props);
  }

  private addDns(api: RestApi, props: MyApiGatewayStackProps) {

    const aRecord = new ARecord(this, 'ApiDNS', {
      recordName: 'api',
      zone: HostedZone.fromHostedZoneAttributes(this, 'HostedZone', {
        hostedZoneId: props.hostedZoneId,
        zoneName: 'thecdkbook.com',
      }),
      target: RecordTarget.fromAlias(new ApiGateway(api)),
    });

    new CfnOutput(this, 'ApiGatewayEndpoint', {
      value: aRecord.domainName,
    });
  }

  private createNewsletterResource(api: RestApi) {
    const newsletterHandler = new NewsletterFunction(this, 'NewsletterHandler');
    const lambdaIntegration = new LambdaIntegration(newsletterHandler);
    const newsletterResource = api.root.addResource('newsletter', {
      defaultCorsPreflightOptions: {
        allowCredentials: true,
        allowHeaders: ['*'],
        allowMethods: ['*'],
        allowOrigins: ['www.thecdkbook.com', 'thecdkbook.com'],
      },
      defaultIntegration: lambdaIntegration,
    });
    newsletterResource.addMethod('GET', lambdaIntegration, {});
  }

  private createReviewsResource(api: RestApi) {
    const reviewsHandler = new ReviewsHandlerFunction(this, 'ReviewsHandler');
    const lambdaIntegration = new LambdaIntegration(reviewsHandler);
    const reviewsResource = api.root.addResource('reviews', {
      defaultCorsPreflightOptions: {
        allowCredentials: true,
        allowHeaders: ['*'],
        allowMethods: ['*'],
        allowOrigins: ['www.thecdkbook.com', 'thecdkbook.com'],
      },
      defaultIntegration: lambdaIntegration,
    });
    reviewsResource.addMethod('GET', lambdaIntegration, {});
  }
}