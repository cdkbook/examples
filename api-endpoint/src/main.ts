import { App } from 'aws-cdk-lib';
import { MyApiGatewayStack } from './MyApiGatewayStack';

const devEnv = {
  account: process.env.CDK_DEFAULT_ACCOUNT,
  region: process.env.CDK_DEFAULT_REGION,
};

const app = new App();

new MyApiGatewayStack(app, 'MyApi', {
  env: devEnv,
  certificateArn: 'arn:aws:acm:us-east-1:536309290949:certificate/92dbf914-21fc-4132-a915-4b67eb029d75',
  hostedZoneId: 'Z09129483SE6UWS2RDQ9Z',
});

app.synth();