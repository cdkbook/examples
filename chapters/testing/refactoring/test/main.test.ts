import { Template } from '@aws-cdk/assertions';
import { App } from '@aws-cdk/core';
import { MyStack } from '../src/main';

test('Snapshot', () => {
  /**
   * This is a simple snapshot test using Jest. If you run this test before your refactoring, it'll take a snapshot
   * of the LogicalId.
   * Refactor your code and then re-run this test. It will fail as the LogicalIds have changed.
   *
   * Then, uncomment out the idMap property and verify the correct values are provided. The failed test will show
   * you the new LogicalId (what it received) and the original LogicalId (the snapshot).
   * In the example code, the new LogicalId of 'ThingyBucket7D8CBF87' will be rewritten to the value it
   * was before the refactor, 'Bucket83908E77', satisfying the test and ensuring your resource won't be recreated.
   */

  const app = new App();
  const stack = new MyStack(app, 'test', { idMap: { ThingyBucket7D8CBF87: 'Bucket83908E77' } });
  expect(Template.fromStack(stack)).toMatchSnapshot();
});