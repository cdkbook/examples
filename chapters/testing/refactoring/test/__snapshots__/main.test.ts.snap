// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot 1`] = `
Object {
  "Resources": Object {
    "Bucket83908E77": Object {
      "DeletionPolicy": "Retain",
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Retain",
    },
  },
}
`;
